dbms/src/Flash/Pipeline/Schedule/Tasks/Task.h

```diff
@@ -17,34 +17,36 @@
 #include <Common/Logger.h>
 #include <Common/MemoryTracker.h>
 #include <Flash/Executor/PipelineExecutorContext.h>
+#include <Flash/Pipeline/Schedule/Tasks/ExecTaskStatus.h>
 #include <Flash/Pipeline/Schedule/Tasks/TaskProfileInfo.h>
 #include <memory.h>
 
 namespace DB
 {
-/**
- *           CANCELLED/ERROR/FINISHED
- *                      ▲
- *                      │
- *         ┌───────────────────────────────┐
- *         │     ┌──►RUNNING◄──┐           │
- * INIT───►│     │             │           │
- *         │     ▼             ▼           │
- *         │ WATITING◄────────►IO_IN/OUT   │
- *         └───────────────────────────────┘
- */
-enum class ExecTaskStatus
+class PipelineExecutorContext;
+
+struct NotifyFuture;
+using NotifyFuturePtr = std::shared_ptr<NotifyFuture>;
+
+struct ReturnStatus
 {
-    WAITING,
-    RUNNING,
-    IO_IN,
-    IO_OUT,
-    FINISHED,
-    ERROR,
-    CANCELLED,
-};
+    ReturnStatus(ExecTaskStatus status_) // NOLINT(google-explicit-constructor)
+        : status(status_)
+        , future(nullptr)
+    {
+        assert(status != ExecTaskStatus::WAIT_FOR_NOTIFY);
+    }
 
-class PipelineExecutorContext;
+    ReturnStatus(NotifyFuturePtr furture_) // NOLINT(google-explicit-constructor)
+        : status(ExecTaskStatus::WAIT_FOR_NOTIFY)
+        , future(std::move(furture_))
+    {
+        assert(future != nullptr);
+    }
+
+    ExecTaskStatus status;
+    NotifyFuturePtr future{nullptr};
```

Comment:
Better to use a `thread_local` variable and do not need to change the return type from `ExecTaskStatus` to `ReturnStatus`.

Reply:
good idea!